domain/cached_repo.py

```diff
@@ -0,0 +1,40 @@
+import logging
+import threading
+from copy import deepcopy
+from datetime import datetime
+
+from domain.model.mesh_results import MeshResults
+from domain.repo import Repo
+
+logger = logging.getLogger(__name__)
+
+
+class CachedRepo:
+    """ CachedRepo allows for thread-safe, periodic updates of the repo data """
+
+    NEVER_UPDATED = datetime(1970, 1, 1)
+
+    def __init__(self, source_repo: Repo, monitored_test_id: str) -> None:
+        self._source_repo = source_repo
+        self._test_id = monitored_test_id
+        self._test_results = MeshResults()
+        self._update_timestamp = CachedRepo.NEVER_UPDATED
+        self._repo_access_lock = threading.Lock()
+
+    def update(self, lookback_minutes: int) -> None:
+        """ update is intended to be called periodically from an update thread """
+
+        try:
+            results = self._source_repo.get_mesh_test_results(self._test_id, lookback_minutes)
+            with self._repo_access_lock:
+                self._test_results = results
+            self._update_timestamp = datetime.now()
```

Comment:
@mateuszmidor I think that `_update_timestamp` should also be synchronized with lock just like `_tests_results`. It is modified in update thread and read in `_make_layout()` call in other threads. Am I missing something?

Reply:
Done